### PR TITLE
Update to Entity Framework 8.0

### PR DESCRIPTION
**Pull request by damieng**
Rebased from #49 to make some additional tweaks.

Thanks to @dylansnel for this work in adjusting the provider to the changes in EF8 that is included in this PR

### CODE REVIEW

**Comment by Temoxa** [comment]
Hi!

How long wait released for EfCore 8.0?

**Comment by damieng** [comment]
Hi.

There won't be another release for a few more months yet - we have a number of features we need to get in first besides the EF8 support such as discriminators, projections, transactions etc.

File: src/MongoDB.EntityFrameworkCore/Storage/MongoTypeMappingSource.cs [hunk 1]
```diff
@@ -76,27 +83,46 @@ public MongoTypeMappingSource(TypeMappingSourceDependencies dependencies)
...
 
         return null;
     }
+
+    public static ValueComparer? CreateComparer(CoreTypeMapping elementMapping, Type collectionType)
+    {
+        // Only single dimension arrays are supported
+        if (collectionType.IsArray && collectionType.GetArrayRank() != 1)
+        {
+            return null;
```
Comment (sanych-sun) [inline src/MongoDB.EntityFrameworkCore/Storage/MongoTypeMappingSource.cs:115]: Should we throw here?

File: src/MongoDB.EntityFrameworkCore/Storage/MongoTypeMappingSource.cs [hunk 2]
```diff
@@ -76,27 +83,46 @@ public MongoTypeMappingSource(TypeMappingSourceDependencies dependencies)
...
+        }
+
+        var elementType = collectionType.TryGetItemType(typeof(IEnumerable<>))!;
+
+        return (ValueComparer?)Activator.CreateInstance(
+            elementType.IsNullableValueType()
+                ? typeof(NullableValueTypeListComparer<>).MakeGenericType(Nullable.GetUnderlyingType(elementType) ?? elementType)
```
Comment (sanych-sun) [inline src/MongoDB.EntityFrameworkCore/Storage/MongoTypeMappingSource.cs:122]: If it takes more then one ternary operator to implement a logic, I prefer to have an good-old `if` instead for better readability.

File: tests/MongoDB.EntityFrameworkCore.FunctionalTests/Mapping/ClrTypeMappingTests.cs [hunk 3]
```diff
@@ -447,11 +448,9 @@ public void Double_negative_read()
     [InlineData(typeof(TestByteEnum?), TestByteEnum.EnumValue1)]
     [InlineData(typeof(int[]), null)]
     [InlineData(typeof(int[]), new[] {-5, 0, 128, 10})]
-    // TODO: investigate and fix IEnumerable property support
-    // [InlineData(typeof(IEnumerable<int>), new[] { -5, 0, 128, 10 })]
     [InlineData(typeof(IList<int>), null)]
     [InlineData(typeof(IList<int>), new[] {-5, 0, 128, 10})]
-    [InlineData(typeof(ICollection<int>), new[] {-5, 0, 128, 10})]
+    [InlineData(typeof(Collection<int>), null)]
```
Comment (sanych-sun) [inline tests/MongoDB.EntityFrameworkCore.FunctionalTests/Mapping/ClrTypeMappingTests.cs:453]: Why do we removed test for the ICollection?

**Review by sanych-sun — CHANGES_REQUESTED** [review]

File: src/MongoDB.EntityFrameworkCore/Storage/MongoTypeMappingSource.cs (hunk 2 above)
Comment (damieng) [reply src/MongoDB.EntityFrameworkCore/Storage/MongoTypeMappingSource.cs:122]: With this approach it's clear that from line 120 onwards it's going to create an instance of something, and that it's dependent on whether it's a nullable value type and if not whether it is assignable or not.

With an if condition you'd have to trawl through 3-4 if statements to find out they all do the same thing with one of three different comparers. It's optimized for scanning through code quickly - the idea being that by the time you've read the first line you can decide whether this is the line you cared about or not - vs reading 15 lines of if conditions only to find they all return a valuecomparer but that wasn't what you were looking for/cared about.

File: tests/MongoDB.EntityFrameworkCore.FunctionalTests/Mapping/ClrTypeMappingTests.cs (hunk 3 above)
Comment (damieng) [reply tests/MongoDB.EntityFrameworkCore.FunctionalTests/Mapping/ClrTypeMappingTests.cs:453]: ICollection<T> is covered by IList<T>.

Collection<T> does not inherit from ICollection<T> and was not previously covered.

File: src/MongoDB.EntityFrameworkCore/Storage/MongoTypeMappingSource.cs (hunk 1 above)
Comment (damieng) [reply src/MongoDB.EntityFrameworkCore/Storage/MongoTypeMappingSource.cs:115]: This block can now actually be removed as the new value comparers support multi-dimensional arrays.

If we don't support that in the C# Driver/BsonDocument then we should add detection to that as part of the model validation instead where we can throw the right error long before anything gets here.

**Comment by damieng** [comment]
Had to force push to deal with merge conflicts cause by progress on main in the mean time.

**Pull request merged**